Adding keypad support to User Lock Manager

blebson opened the pull request:
I've added basic keypad support to your most recent version of User Lock Manager. It's still a bit of a hack job but I currently have the following features added/working:

-Associate Multiple Keypads (not yet tested with multiple)
-Allow keypads to use existing codes
-Report who armed/disarmed
-Add to code usage count when arming/disarming keypad
-Select which Routine you want to run for each arm/disarm mode

This is currently confirmed to be working with the Centralite 3400-X keypad (made for Xfinity). May also work with Lowes Iris keypad, though I haven't tested it.

blebson committed:
Update and rename user-lock-manager.smartapp.groovy to user-lock-manager-w-keypad-support.smartapp.groovy

initial release of user lock manager with keypad support.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

fixed keypad support

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

Fixed keypad support, added delay

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

added arm/disarm message and code usage incrementation

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

fixed code usage count.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

Fixed code usage reporting

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

cleaned up code

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

re-enabled 'Night Mode'

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

added burn code functionality to keypad entered codes

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

Fixed 'One Time Use' codes.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

Fixed one time use code.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

Made it work better when using multiple instances of the SmartApp simultaneously.

ethayer commented on user-lock-manager-w-keypad-support.smartapp.groovy:
```diff
@@ -69,7 +72,7 @@ def setupPage() {
           }
           if (settings."userCode${user}" && settings."userSlot${user}") {
-            getConflicts(settings."userSlot${user}")
+            //getConflicts(settings."userSlot${user}")
```
What are these comments doing around?

blebson replied:
I had to comment out the calls to getConflicts due to it preventing me from accessing the user settings in a second instance of the SmartApp. It was giving me a 'you are not authorized...' error message.

ethayer commented on user-lock-manager-w-keypad-support.smartapp.groovy:
```diff
@@ -96,16 +99,16 @@ def userPage(params) {
+      //      }
+      //    }
+      //  }
+      //}
```
Commented out code?

blebson replied:
This goes back to the error caused by getConflicts which prevented me from viewing user info in a second instance of the SmartApp.

ethayer commented on user-lock-manager-w-keypad-support.smartapp.groovy:
```diff
@@ -1066,20 +1092,29 @@ def performActions(evt) {
-        }
-      } else {
-       location.helloHome.execute(homePhrases)
+      if(location.currentMode != homePhrases)
```
Would you mind keeping with the code style of the rest of the document?  example:

```
if (something) {
 //do some things
}
```

instead of:

```
if (something)
{
  //do something else
}
```

blebson replied:
I apologize for any formatting sloppiness, I unfortunately don't have any code editing program (i.e. Visual Studio, Eclipse, etc.) so this is all edited in the IDE directly. I'll work to clean it up.

blebson replied:
Currently this check isn't working properly. The idea is to prevent the SmartApp from running a Routine if it's already in the mode that it's trying to change to. This is an issue when you have multiple users without presence devices and they keep adjusting the thermostat every time someone unlocks a door.

ethayer commented:
Awesome work.  I don't have a pin pad to test it out, but I'm thinking one might be cool.  Do you have one you'd suggest?

Fix the issues and I'll merge this in.

blebson commented:
You can either get the CentraLite 3400-X off of eBay (it's generally sold as an Xfinity Home accessory) or you can get the new Lowes Iris Keypad (purple box). They're both CentraLite models and can use the same devicetype. These are the only ones that I'm aware of that work right now. I personally have the CentraLite 3400-X which I think looks better than the Lowes Iris version.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

Updated format to match existing code

blebson commented:
I fixed the formatting issues. The issues that are still outstanding are as follows:

getConflicts is still commented out, not sure what to do in order to resolve the 'You are not authorized...' errors. I can un-comment the code but am worried about how to modify users in the second instance if it keeps throwing the error.

Line 1095, code to prevent running of Routine if already in the requested mode is not currently working. Not sure how to get which mode it's trying to change into from a routine.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

fixed Routine ignore when in specific mode.

blebson commented:
latest update should properly add prevention of running Routine on unlock if in a specific mode. Added variable to 'Schedule' section to allow user to select mode they would like to ignore Routine in.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

removed comments from getConflicts code. Modified message sent if routine change is ignored due to being in request mode already.

blebson commented:
I believe I made all the requested changes. Still no resolution on fixing the getConflicts errors but I removed the comments from the code.

Let me know if there are other changes that need to happen before you can merge.

blebson commented:
This is the error I get when trying to modify a user in my second instance of the SmartApp (1 user w/ code). I have two instances so that I can have one code on a schedule while the others are not. This error does not affect the first instance of the SmartApp (4 users w/ codes).

![image](https://cloud.githubusercontent.com/assets/14287604/11129770/12997588-8936-11e5-8465-81e2ef47f0fb.png)

ctennant85 commented:
Hi I'm a landlord of student share property, I'm planning on buying the Yale lock in the link below. I would like to designate a code to each student valid for a date range and generate a temporary code for tradesmen. I would like to does this remotely. Can you tell me if this would be possible with the program you have written and if not can I pay you to make it work? You can contact me via email ctennant@me.com

Cheers Chris

http://www.yalestore.co.uk/smart-living/keyfree-connected-lock.html

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy

Updated base code to 4.1.0

blebson commented:
I just updated the base code to 4.1.0, I'm not really familiar with how pull requests work but it's still comparing it to the 4.0.7 base code....

ethayer commented on user-lock-manager-w-keypad-support.smartapp.groovy:
```diff
@@ -1293,3 +1340,154 @@ def populateDiscovery(codeData, lock) {
+        def message = " "
+        
+        i = settings.maxUsers
+    while (i > 0) {
```
Wouldn't this just stop at 1 execution?  Do you mean to do something like...

```
for (int i = 1; i <= settings.maxUsers; i++) {
```

blebson replied:
this section is just looping through once for each user in the app. It starts at the max number and counts down until it hits 0. When it matches a code with the code entered it sets the variable to 0. I'm not familiar with the 'for' loop which is why I just used the 'while'. It works properly as it is right now.

ethayer commented on user-lock-manager-w-keypad-support.smartapp.groovy:
```diff
@@ -682,6 +711,8 @@ private initialize() {
   subscribe(theLocks, "lock", codeUsed)
   subscribe(theLocks, "reportAllCodes", pollCodeReport, [filterEvents:false])
+  subscribe(location,"alarmSystemStatus",alarmStatusHandler)
+  subscribe(keypad,"codeEntered",codeEntryHandler)
```
Maybe wrap these if an if block for performance reasons...

```
if (keypads) {
  subscribe(location,"alarmSystemStatus",alarmStatusHandler)
  subscribe(keypad,"codeEntered",codeEntryHandler)
}
```

blebson replied:
I could definitely do that, what would be the best way to determine if the keypad is enabled/set?

ethayer replied:
`if (keypads) {}`

works well.  Or keypad, but since you're allowing multiple keypads I'd call the variable 'keypads'.

if no keypads are set, keypads will be falsely and the block won't run.

blebson replied:
Ok, I just updated the code to add this logic.

ethayer commented on user-lock-manager-w-keypad-support.smartapp.groovy:
```diff
@@ -48,6 +50,7 @@ def rootPage() {
         input name: "maxUsers", title: "Number of users", type: "number", multiple: false, refreshAfterSelection: true, submitOnChange: true
         href(name: "toSetupPage", title: "User Settings", page: "setupPage", description: setupPageDescription(), state: setupPageDescription() ? "complete" : "")
         href(name: "toInfoPage", page: "infoPage", title: "Lock Info")
+        href(name: "toKeypadPage", page: "keypadPage", title: "Keypad Info (optional)")
```
Wrap this in an if block too..
if (keypads) {
  href(name: "toKeypadPage", page: "keypadPage", title: "Keypad Info (optional)")
}

ethayer commented on user-lock-manager-w-keypad-support.smartapp.groovy:
```diff
@@ -1293,3 +1340,154 @@ def populateDiscovery(codeData, lock) {
+    	}
+    }
+    changedMode = 1
+    i--
```
Ah, I see it now.  Our minds are just going in opposite directions.

```
for (int i = 1; i <= settings.maxUsers; i++) {
}
```

Is the same thing, but sets the variable and INCREMENTS the index through the loop.  I think this is easier to read since the logic is all in one place, but either one works.  Carry on.

blebson committed:
Update user-lock-manager-w-keypad-support.smartapp.groovy
